spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestChangelogReader.java

```diff
@@ -191,39 +214,361 @@ public void testMixDeleteAndInsert() throws IOException {
+
+    List<Object[]> expectedRows = Lists.newArrayList();
+    addExpectedRows(expectedRows, ChangelogOperation.INSERT, snapshotId1, 0, rl1);
+    addExpectedRows(expectedRows, ChangelogOperation.DELETE, snapshotId2, 1, deleted1);
+    addExpectedRows(expectedRows, ChangelogOperation.INSERT, snapshotId2, 1, rl2);
+    // in snapshot 3, eqDeletes2 applies to both df1 and df2, so there are two deletes emitted
+    addExpectedRows(expectedRows, ChangelogOperation.DELETE, snapshotId3, 2, deleted1);
+    addExpectedRows(expectedRows, ChangelogOperation.DELETE, snapshotId3, 2, deleted2);
+    addExpectedRows(expectedRows, ChangelogOperation.INSERT, snapshotId3, 2, rl3);
+
+    assertEquals("Should have expected rows", expectedRows, internalRowsToJava(rows));
+  }
+
+  @Test
+  public void testFlinkScenario2() throws IOException {
+    List<Record> rl1 = Lists.newArrayList();
+    GenericRecord record = GenericRecord.create(table2.schema());
+    rl1.add(record.copy("id", 1, "data", "a"));
+    rl1.add(record.copy("id", 2, "data", "b"));
+    rl1.add(record.copy("id", 3, "data", "c"));
+    DataFile df1 = writeDataFile(rl1);
+    table2.newAppend().appendFile(df1).commit();
+    long snapshotId1 = table2.currentSnapshot().snapshotId();
+
+    // delete row with id=2 with an equality delete and write a new row for it, but then delete the
+    // new row with a positional delete and write yet another new row for it
+    Schema deleteRowSchema = table2.schema().select("id");
+    Record idDelete = GenericRecord.create(deleteRowSchema);
+    List<Record> idDeletes = Lists.newArrayList(idDelete.copy("id", 2));
+
+    DeleteFile eqDeletes =
+        FileHelpers.writeDeleteFile(
+            table2,
+            Files.localOutput(File.createTempFile("junit", null, temp.toFile())),
+            TestHelpers.Row.of(0),
+            idDeletes,
+            deleteRowSchema);
+
+    List<Record> rl2 = Lists.newArrayList();
+    rl2.add(record.copy("id", 2, "data", "bb"));
+    rl2.add(record.copy("id", 4, "data", "d"));
+    rl2.add(record.copy("id", 5, "data", "e"));
+    rl2.add(record.copy("id", 2, "data", "bbb"));
+    DataFile df2 = writeDataFile(rl2);
+
+    List<Pair<CharSequence, Long>> deletes = Lists.newArrayList(Pair.of(df2.path(), 0L));
+
+    Pair<DeleteFile, CharSequenceSet> posDeletes =
+        FileHelpers.writeDeleteFile(
+            table,
+            Files.localOutput(File.createTempFile("junit", null, temp.toFile())),
+            TestHelpers.Row.of(0),
+            deletes);
+
+    table2
+        .newRowDelta()
+        .addDeletes(eqDeletes)
+        .addRows(df2)
+        .addDeletes(posDeletes.first())
+        .validateDataFilesExist(posDeletes.second())
+        .commit();
+    long snapshotId2 = table2.currentSnapshot().snapshotId();
+
+    List<InternalRow> rows = getChangelogRows(table2);
+
+    // for snapshot 2, we should record only one delete for id=2 (from the equality delete)
+    // and one insert (the final value)
```

Comment:
This is one way to present this data. And I would be totally fine with this result. For Flink this is as good as it can get.

Other writers might write delete records into the positional delete records (like `df1.path(), 0L` - to remove "1,a") along with the ones which remove the the new records. If those are handled correctly, then we are golden here.